python/lib/core/dmod/core/dataset.py

```diff
@@ -487,6 +489,360 @@ def unlink_to_dataset(self, dataset: Dataset) -> bool:
+        self._data_collection: DataCollection = data_collection
+        """ Collection of data items, analogous to a :class:`Dataset`, if not a dataset outright. """
+        self._collection_name: Optional[str] = collection_name
+        """
+        Optional name for collection, which is important when domains involve a ``data_id`` restriction.
+        
+        Note that **IFF** the data collection is a :class:`Dataset` object **AND** the passed ``collection_name`` param
+        is ``None``, then this attribute will be set to the ``name`` attribute of the dataset collection itself.
+        """
+        if collection_name is None and isinstance(data_collection, Dataset):
+            self._collection_name = data_collection.name
+
+        if isinstance(self._data_collection, Path) and not self._data_collection.is_dir():
+            raise ValueError(f"{self.__class__.__name__} initialized with a path require this path to be a directory.")
+        if isinstance(self._data_collection, Dataset) and self._data_collection.manager is None:
+            raise ValueError(f"Dataset used to initialize {self.__class__.__name__} must have manager set.")
+
+    def detect(self, **kwargs) -> DataDomain:
+        """
+        Detect and return the data domain.
+
+        Parameters
+        ----------
+        kwargs
+            Optional kwargs applicable to the subtype, which may enhance or add to the domain detection and generation
+            capabilities, but which should not be required to produce a valid domain.
+
+        Notes
+        -----
+        Detection is performed by merging individual item domains detected using :class:`UniversalItemDomainDetector`
+        instances.  This type does not influence the details of how individual domains detections are performed by
+        :class:`UniversalItemDomainDetector` objects.  The subsequent merging is performed by reducing the
+        individual item domains using :method:`DataDomain.merge_domains`. The order of the items processed when reducing
+        is based on the order of results of a call to :method:`get_item_names`.
+
+        Returns
+        -------
+        DataDomain
+            The detected domain.
+
+        Raises
+        ------
+        DmodRuntimeError
+            If it was not possible to properly detect the domain.
+        """
+        domain = reduce(DataDomain.merge_domains,
+                        {UniversalItemDomainDetector(item=self.get_item(i), item_name=i).detect() for i in
+                         self.get_item_names()})
+        # If this domain has a format with a self-reference to dataset id, and we have a name, then set that restriction
+        if StandardDatasetIndex.DATA_ID in domain.data_format.indices_to_fields().keys() and self._collection_name:
+            domain.discrete_restrictions[StandardDatasetIndex.DATA_ID] = DiscreteRestriction(
+                variable=StandardDatasetIndex.DATA_ID, values=[self._collection_name])
```

Comment:
Hmmm ... I used an IDE feature to automatically do this reformatting, which I assume uses standard styling.   I don't think this counts as having the parameter list spanning multiple lines.   I played around with things a little, and it seems to be because there wasn't enough room on the first line for even the first param.  If I refactor things so that all of `variable` fits on the first line (but not `values`), the IDE reformats in the manner suggested:

```Python
if StandardDatasetIndex.DATA_ID in domain.data_format.indices_to_fields().keys() and self._collection_name:
            standard_index = StandardDatasetIndex.DATA_ID
            domain.discrete_restrictions[standard_index] = DiscreteRestriction(variable=standard_index,
                                                                               values=[self._collection_name])
```

So, how do we want the DMOD style to handle this?

Reply:
I would go:

```python
format_contains_index = StandardDatasetIndex.DATA_ID in domain.data_format.indices_to_fields().keys()
if format_contains_index and self._collection_name:
    standard_index = StandardDatasetIndex.DATA_ID
            
    domain.discrete_restrictions[standard_index] = DiscreteRestriction(
        variable=standard_index,
        values=[self._collection_name]
    )
```

or

```python
format_contains_index = StandardDatasetIndex.DATA_ID in domain.data_format.indices_to_fields().keys()
if format_contains_index and self._collection_name:
    standard_index = StandardDatasetIndex.DATA_ID
    restriction = DiscreteRestriction(variable=standard_index, values=[self._collection_name])        
    domain.discrete_restrictions[standard_index] = restriction
```

If you have a long invocation that goes past the end of the screen, separating the input parameters on individual lines makes it easy to differentiate one value from another and the single lines make it easier to see changes within diff tools, which is pretty nice.

Your example looks fine, though I'm not its biggest fan - that might just be because I prefer the first parameter to be on its own as well to help create that visual barrier, but that might just be personal preference.